### ISSUE TITLE
Misaligned module code in Prerequisite Tree

### ISSUE THREAD

**Issue by Jiwei99** [nusmodifications/nusmods#3841]
### Describe the bug

Some module code in the Pre-Requisite Tree is misaligned.

### To Reproduce

<img width="962" alt="Screenshot 2024-10-22 at 10 46 16 AM" src="https://github.com/user-attachments/assets/319f2cb7-87c6-4160-bfda-fa17011066d0">
Layout of FIN2704 is different from rest of the modules <br>
https://nusmods.com/courses/FIN3716/financial-modelling
<br><br>
<img width="1035" alt="Screenshot 2024-10-22 at 10 48 42 AM" src="https://github.com/user-attachments/assets/ecbb74da-277d-4ef5-9fb1-84e60d46cdda">
Module codes are not aligned properly, and the alignment is inconsistent <br>
https://nusmods.com/courses/BSP3701A/strategic-management

### Desktop (please complete the following information):

- OS: [MacOS]
- Browser [Chrome, Safari]

**Comment by zehata** [comment]
Looks like it's caused by FIN2704 being an available module, which wraps it in a button that is invisible
![image](https://github.com/user-attachments/assets/8228a00c-a69a-4673-a904-05e5a1e866ef)

**Comment by zehata** [comment]
FIN2704 is an active mod that links to FIN2704/FIN2704X, so `getModuleCondensed("FIN2704")` is truthy but `moduleActive` for FIN2704 is falsey (name is not exact match). This causes the node to be a clickable button that appears invisible.

Fix is trivial but UX requires consideration, there are a few options:
1. Remove the button, keep the text. Preferred for consistency.
![image](https://github.com/user-attachments/assets/c4ecf048-3783-48ed-bfff-c03f1ab79dda)
2. Keep the button, and make it visible
![image](https://github.com/user-attachments/assets/8da5de54-9c81-4a0b-bf99-2385cfd2813c)
3. Keep the button, keep it invisible(??) Really strange UI
![image](https://github.com/user-attachments/assets/8aa3578f-49c3-4590-8934-567879376711)

**Comment by Jiwei99** [comment]
@zehata What if "Course starting with" is replaced with an enumeration of all versions of the module using a "one of" branch? Think that would be more aligned with the design of the rest of the prerequisites trees (but not sure how feasible is this)

**Comment by zehata** [comment]
Alternatively:
![image](https://github.com/user-attachments/assets/7a1d4cd5-ba4b-4a43-9ea9-9725a4f240b3)

**Comment by Jiwei99** [comment]
> Alternatively: ![image](https://private-user-images.githubusercontent.com/16816749/379795812-2c70b119-8a6a-4631-b087-efc68795523f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3ODA2NDcsIm5iZiI6MTcyOTc4MDM0NywicGF0aCI6Ii8xNjgxNjc0OS8zNzk3OTU4MTItMmM3MGIxMTktOGE2YS00NjMxLWIwODctZWZjNjg3OTU1MjNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI0VDE0MzIyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkODNhZTMzNGFlYmU2N2QwZjY1MDM3NzdlNzA1MDgyOWRkZGVkNmVmMzhjMTE0YjUzNThmOTFkODJmZGNkNTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lUkGaAQwVF_xMFP6mMzY3MyrSNYzmXNai9OnajyIvD0)

Yeah I think this looks the cleanest, but should the grade requirement be shown? I don’t remember seeing it in the current prereq tree

**Comment by zehata** [comment]
The grade requirement is only shown when it is not the default 'D' grade.

**Comment by Jiwei99** [comment]
> Alternatively: ![image](https://private-user-images.githubusercontent.com/16816749/379795812-2c70b119-8a6a-4631-b087-efc68795523f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3ODcxNzEsIm5iZiI6MTcyOTc4Njg3MSwicGF0aCI6Ii8xNjgxNjc0OS8zNzk3OTU4MTItMmM3MGIxMTktOGE2YS00NjMxLWIwODctZWZjNjg3OTU1MjNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI0VDE2MjExMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjNDFjN2Y0MTg5MWM0YjIxZWEzNWM5ZTYzMWMzY2VjY2E1YWU3MGM5ZGY5ZmZlMTZiZWUwZDNiNTUwZGRkOWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ULvMO4JXMGhtWmrDE9KjHdBe0uMuBHP5Gpn_lRaZ7N8)

Ah, then I would think this looks the best. The other version with the prefix in the module button looks too chunky

**Comment by zehata** [comment]
It is possible to do this, but it's possible that situations like this will happen, which might appear messy to some people? (I think it's fine though). If it's possible, I hope someone from the core team can weigh in on this UI. (The tree below is just test data) 
![image](https://github.com/user-attachments/assets/c13b9162-6b54-4f69-838a-7b6bad4af6df)

**Comment by zehata** [comment]
Also @ core team, an alternative is to update ANTLR to branch when encountering wildcard modules instead of appending % when scraping, which is a cleaner implementation but will also be more complex computationally on the server. I think practically speaking this frontend implementation is sufficient.

**Comment by ravern** [comment]
> but will also be more complex computationally on the server.

I think this is fine -- we only run the scraper once every hour.

I can see this working by modifying this `parse` function to build up a trie of modules (opposed to a list) and then passing that to the `parseString` function, which can then expand modules it comes across.  Alternatively we need a new function `expandWildcards` after the `parseString` function call.

https://github.com/nusmodifications/nusmods/blob/0f33a55fb2c446298016be8e8d499c030e099c4d/scrapers/nus-v2/src/services/requisite-tree/index.ts#L27-L48

--------------------------

But going back to earlier discussion, do we want to expand these "courses starting with" in the first place? I'm afraid that it might generate huge and unwieldy pre-req trees in some cases.

If anyone has some time, could we perhaps get a sense of what's the largest pre-req tree this could potentially generate (in terms of number of rows/lines)? Probably by writing a quick script that runs across all the modules in `moduleList.json` and expands everything.

**Comment by zehata** [comment]
@ravern Sorry about the long silence, I wasn't paying attention to this because exam season and also my github notifs aren't working for some reason...

**Comment by zwliew** [comment]
I don't really have an opinion on this for now, but I'm going to first merge a stopgap solution for the misalignment issue (#3883). We can work on further enhancements to the UI later on.
